bee-network/bee-gossip/src/handshake.rs

```diff
@@ -0,0 +1,380 @@
+// Copyright 2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    conn::Direction,
+    consts::{HANDSHAKE_TIMEOUT_SECS, MAX_HANDSHAKE_PACKET_SIZE, VERSION},
+    proto,
+};
+
+use bee_identity::identity::{LocalId, PeerId};
+use bee_network_packet::{packet_hash, Packet, PacketType};
+use bee_peering_manual::config::ManualPeerConfig;
+
+use crypto::signatures::ed25519;
+use prost::{bytes::BytesMut, Message as _};
+use tokio::{
+    io::{self, AsyncReadExt, AsyncWriteExt, BufReader, BufWriter},
+    net::{
+        tcp::{OwnedReadHalf, OwnedWriteHalf},
+        TcpStream,
+    },
+};
+
+use std::{
+    fmt,
+    net::{IpAddr, SocketAddr},
+    time::{SystemTime, UNIX_EPOCH},
+};
+
+type Alias = String;
+
+pub async fn handshake(
+    stream: TcpStream,
+    socket_addr: SocketAddr,
+    local_id: &LocalId,
+    direction: Direction,
```

Comment:
I wonder if this argument could be a `const OUTBOUND: bool` generic parameter. What do you think @Alex6323?

Reply:
I did the change :)

Reply:
We can do that @pvdrz , but it comes (I think) with two downsides (that we need to decide it's worth having them):
(1) `match handshake::<false>(....)` is a bit less readable:
The reader has to first look up `handshake`s signature to understand what the consequences of `false` are. (Not really a problem for most IDE users which usually will pop up the functions documentation.)
(2) Increased binary size. Monomorphization will duplicate this function as they are both used.

What are your thoughts on it? @pvdrz

Reply:
I think (2) is negligible in this case as the function is not that big, but (1) disturbs me quite a bit. If it would be `handshake::<Direction::Inbound>(...) things were different.

Reply:
Yeah the cost of (2) doesn't worry me that much because we end up with two shorter functions instead of a longer one (even though the two functions are longer together).

About (1) it is a bit hard because const generics only allows bools and ints for now as parameters. We could have two constants called `OUTBOUND` and `INBOUND` instead. What do you think?